server/src/main/java/com/cloud/vm/UserVmManagerImpl.java

```diff
@@ -2335,6 +2337,15 @@ public UserVm recoverVirtualMachine(RecoverVMCmd cmd) throws ResourceAllocationE
         return _vmDao.findById(vmId);
     }
 
+    protected void recoverRootVolume(VolumeVO volume, Long vmId) {
+        if (Volume.State.Destroy.equals(volume.getState())) {
+            _volumeService.recoverVolume(volume.getId());
+            _volsDao.attachVolume(volume.getId(), vmId, ROOT_DEVICE_ID);
+        } else {
+            _volumeService.publishVolumeCreationUsageEvent(volume);
+        }
```

Comment:
doesn´t this mean a volume creation event gets published when the volume is in a state of Expunging/Expunged?

Reply:
This is current behavior, I just added the case where the root volume is destroyed. If you look at the current code, no verification is made. An event is always published.